File: src/lua.ts
/**
 * Parse openlua.cloud-style SteamTools lua files.
 *
 * Handles:
 *   - Zero-width watermark chars (openlua.cloud fingerprints leaked files)
 *   - addappid(APPID, 1, "DEPOT_KEY")
 *   - setManifestid(DEPOTID, "MANIFEST_ID", 0)
 */

export interface DepotEntry {
  id: number;
  key: string;          // hex
  manifestId?: string;  // present only if setManifestid was used
  label?: string;       // from trailing comment, best-effort
}

export interface LuaParseResult {
  appId: number;        // main app id (first addappid)
  depots: DepotEntry[]; // includes main app, content depots, DLCs
  raw: string;          // cleaned source
}

const ZERO_WIDTH = /[\u200B-\u200D\uFEFF\u2060]/g;

export function parseLua(source: string): LuaParseResult {
  const clean = source.replace(ZERO_WIDTH, "");

  const depots = new Map<number, DepotEntry>();

  // addappid(APPID, 1, "KEY")  — optional trailing comment
  const addRe =
    /addappid\s*\(\s*(\d+)\s*,\s*\d+\s*,\s*"([0-9a-fA-F]+)"\s*\)\s*(?:--\s*(.*))?/g;
  for (const m of clean.matchAll(addRe)) {
    const id = Number(m[1]);
    depots.set(id, {
      id,
      key: m[2].toLowerCase(),
      label: m[3]?.trim(),
    });
  }

  // setManifestid(DEPOTID, "MID", 0)
  const setRe = /setManifestid\s*\(\s*(\d+)\s*,\s*"(\d+)"/g;
  for (const m of clean.matchAll(setRe)) {
    const id = Number(m[1]);
    const existing = depots.get(id);
    if (existing) existing.manifestId = m[2];
    else depots.set(id, { id, key: "", manifestId: m[2] });
  }

  const first = [...depots.values()].find((d) => d.key);
  if (!first) throw new Error("No addappid(...) entries found");

  return {
    appId: first.id,
    depots: [...depots.values()],
    raw: clean,
  };
}

File: internal/lua/lua.go
// Package lua parses openlua.cloud-style SteamTools lua files.
//
// Handles:
//   - Zero-width watermark chars (openlua.cloud fingerprints leaked files)
//   - addappid(APPID, 1, "DEPOT_KEY")
//   - setManifestid(DEPOTID, "MANIFEST_ID", 0)
package lua

import (
	"errors"
	"regexp"
	"strconv"
	"strings"
)

type DepotEntry struct {
	ID         uint32
	Key        string // lowercase hex
	ManifestID string // present only if setManifestid was used
	Label      string // from trailing comment, best-effort
}

type ParseResult struct {
	AppID  uint32
	Depots []DepotEntry
	Raw    string
}

var (
	zeroWidth = regexp.MustCompile("[\u200B-\u200D\uFEFF\u2060]")
	addRe     = regexp.MustCompile(`addappid\s*\(\s*(\d+)\s*,\s*\d+\s*,\s*"([0-9a-fA-F]+)"\s*\)[ \t]*(?:--[ \t]*([^\n]*))?`)
	setRe     = regexp.MustCompile(`setManifestid\s*\(\s*(\d+)\s*,\s*"(\d+)"`)
)

func Parse(source string) (*ParseResult, error) {
	clean := zeroWidth.ReplaceAllString(source, "")

	// Ordered map: insertion order matters for reproducibility.
	type entry struct {
		idx int
		e   DepotEntry
	}
	depots := make(map[uint32]*entry)
	order := 0

	for _, m := range addRe.FindAllStringSubmatch(clean, -1) {
		id, err := strconv.ParseUint(m[1], 10, 32)
		if err != nil {
			continue
		}
		key := strings.ToLower(m[2])
		label := ""
		if len(m) > 3 {
			label = strings.TrimSpace(m[3])
		}
		depots[uint32(id)] = &entry{
			idx: order,
			e:   DepotEntry{ID: uint32(id), Key: key, Label: label},
		}
		order++
	}

	for _, m := range setRe.FindAllStringSubmatch(clean, -1) {
		id, err := strconv.ParseUint(m[1], 10, 32)
		if err != nil {
			continue
		}
		if existing, ok := depots[uint32(id)]; ok {
			existing.e.ManifestID = m[2]
		} else {
			depots[uint32(id)] = &entry{
				idx: order,
				e:   DepotEntry{ID: uint32(id), ManifestID: m[2]},
			}
			order++
		}
	}

	// Flatten preserving insertion order.
	out := make([]DepotEntry, len(depots))
	for _, e := range depots {
		out[e.idx] = e.e
	}
	// Compact in case of gaps (should not occur with single-threaded parse).
	compact := out[:0]
	for _, e := range out {
		if e.ID != 0 {
			compact = append(compact, e)
		}
	}

	// First entry with a key is the "main" app.
	var appID uint32
	for _, d := range compact {
		if d.Key != "" {
			appID = d.ID
			break
		}
	}
	if appID == 0 {
		return nil, errors.New("no addappid(...) entries found")
	}

	return &ParseResult{AppID: appID, Depots: compact, Raw: clean}, nil
}
